modules/codebuild/main.tf

```diff
@@ -0,0 +1,523 @@
+        "logs:CreateLogGroup",
+        "logs:CreateLogStream",
+        "logs:PutLogEvents",
+
+        # VPC permissions
+        "ec2:CreateNetworkInterface",
+        "ec2:DescribeDhcpOptions",
+        "ec2:DescribeNetworkInterfaces",
+        "ec2:DeleteNetworkInterface",
+        "ec2:DescribeSubnets",
+        "ec2:DescribeSecurityGroups",
+        "ec2:DescribeVpcs",
+
+        # ECR permissions
+        "ecr:GetAuthorizationToken",
+        "ecr:BatchCheckLayerAvailability",
+        "ecr:GetDownloadUrlForLayer",
+        "ecr:GetRepositoryPolicy",
+        "ecr:DescribeRepositories",
+        "ecr:ListImages",
+        "ecr:DescribeImages",
+        "ecr:BatchGetImage",
+        "ecr:GetLifecyclePolicy",
+        "ecr:GetLifecyclePolicyPreview",
+        "ecr:ListTagsForResource",
+        "ecr:DescribeImageScanFindings",
+
+        # SSM Parameter Store
+        "ssm:DeleteParameter",
+        "ssm:DeleteParameters",
+        "ssm:DescribeParameters",
+        "ssm:GetParameter",
+        "ssm:GetParameters",
+        "ssm:GetParameterHistory",
+        "ssm:GetParametersByPath",
+        "ssm:PutParameter",
+
+        # Secrets Manager
+        "secretsmanager:GetResourcePolicy",
+        "secretsmanager:GetSecretValue",
+        "secretsmanager:DescribeSecret",
+        "secretsmanager:ListSecrets",
+        "secretsmanager:ListSecretVersionIds"
+      ], var.extra_permissions))
+
+      effect = "Allow"
+
+      resources = [
+        "*",
+      ]
+    }
+  }
+
+  dynamic "statement" {
+    for_each = var.secondary_artifact_location != null ? [1] : []
+    content {
+      sid = ""
+
+      actions = [
+        "s3:PutObject",
+        "s3:GetBucketAcl",
+        "s3:GetBucketLocation"
+      ]
+
+      effect = "Allow"
+
+      resources = [
+        join("", data.aws_s3_bucket.secondary_artifact[*].arn),
+        "${join("", data.aws_s3_bucket.secondary_artifact[*].arn)}/*",
+      ]
+    }
+  }
+
+  lifecycle {
+    precondition {
+      condition     = length(var.extra_permissions) > 0 ? var.default_permissions_enabled : true
+      error_message = <<-EOT
+      Extra permissions can only be attached to the default permissions policy statement.
+      Either set `default_permissions_enabled` to true or use `custom_policy` to set a least privileged policy."
+      EOT
+    }
+  }
+}
+
+data "aws_iam_policy_document" "vpc_permissions" {
+  count = var.vpc_config != {} ? 1 : 0
+
+  statement {
+    sid = ""
+
+    actions = [
+      "ec2:CreateNetworkInterface",
+      "ec2:DescribeDhcpOptions",
+      "ec2:DescribeNetworkInterfaces",
+      "ec2:DeleteNetworkInterface",
+      "ec2:DescribeSubnets",
+      "ec2:DescribeSecurityGroups",
+      "ec2:DescribeVpcs"
+    ]
+
+    resources = [
+      "*",
+    ]
+  }
+
+  statement {
+    sid = ""
+
+    actions = [
+      "ec2:CreateNetworkInterfacePermission"
+    ]
+
+    resources = [
+      "arn:aws:ec2:${local.aws_region}:${local.aws_account_id}:network-interface/*"
```

Comment:
i think these will need to be pulled from a variable or data object - 

```
data "aws_partition" "current" {
}
```

since the partition could also be `arn:aws-us-gov:` in govcloud